src/herder/HerderImpl.cpp

```diff
@@ -148,19 +148,37 @@ void
 HerderImpl::valueExternalized(uint64 slotIndex, StellarValue const& value)
 {
     const int DUMP_SCP_TIMEOUT_SECONDS = 20;
+    bool newSlot = slotIndex == getCurrentLedgerSeq();
 
-    // record metrics
-    getHerderSCPDriver().recordSCPExecutionMetrics(slotIndex);
+    if (newSlot)
+    {
+        getHerderSCPDriver().recordSCPExecutionMetrics(slotIndex);
+        // dump SCP information if this ledger took a long time
+        auto gap = std::chrono::duration<double>(mApp.getClock().now() -
+                                                 mLastExternalize)
+                       .count();
+        if (gap > DUMP_SCP_TIMEOUT_SECONDS)
+        {
+            auto slotInfo = getJsonQuorumInfo(getSCP().getLocalNodeID(), false,
+                                              false, slotIndex);
+            Json::FastWriter fw;
+            CLOG(WARNING, "Herder")
+                << fmt::format("Ledger took {} seconds, SCP information:{}",
+                               gap, fw.write(slotInfo));
+        }
+    }
```

Comment:
Wouldn't it be easier to refactor this code instead of adding a bunch of `if(newSlot)`?

Right now you're still including certain things that should not be included when processing an old slot:
* calls to `trackingHeartBeat()`
* `mTriggerTimer.cancel()`
* `checkAndMaybeReanalyzeQuorumMap()`
* `ledgerClosed()`

basically you're left with only the "central part" of this function that does:
* logging
* saving of SCP message
* cleanup upgrades
* call ledger manager's `valueExternalized`